Extract StaticContentType from StaticContentAdmin

fbourigault opened the pull request:
To ease the extraction of sonata admin class, it was decided that the form used to edit staticContent should be extracted from the admin class.

I also changed a bit options about IvoryCkEditor to allow using IvoryCkEditor with default options.

## Progress

### Todo
- [x] parentDocument field
- [x] TreeSelectType assets
- [x] allow to use TreeSelectType browser widget
- [x] fix fields ordering
- [x] tests
- [x] history rewrite
- [x] write `Form/Extension/CKEditorExtension` tests
- [x] squash commits

### Blockers
- [x] https://github.com/symfony-cmf/tree-browser-bundle/pull/112
- [x] https://github.com/sonata-project/SonataAdminBundle/pull/4101

wouterj commented on Admin/StaticContentAdmin.php:
```diff
@@ -49,18 +46,13 @@ protected function configureFormFields(FormMapper $formMapper)
-                    $this->ivoryCkeditor
-                )
+                ->add('static_content', StaticContentType::class, [
+                    'inherit_data' => true,
```
let's configure this one in the StaticContentType, I can't imagine a case where we use StaticContentType without wanting it to inherit_data.

fbourigault replied:
I put `inherit_data` in the form because it may break the form if someone wants to use it through `Symfony\Bundle\FrameworkBundle\Controller::createForm`. I'm not sure if it breaks or not, but it could be a legitimate use of the `StaticContentType`.

wouterj replied:
Oh indeed, makes sense now

wouterj commented on Form/Type/StaticContentType.php:
```diff
@@ -0,0 +1,43 @@
+    {
+        $resolver->setDefaults([
+            'data_class' => StaticContent::class,
+            'phpcr_rootpath' => '',
```
I think a `/` default root path makes more sense

fbourigault commented:
I looked at the `parentDocument` field about missing options. As you told me those options are added by the FormContractor which is tightly coupled with the admin class. Should I add two options to the `StaticContentType` to emulate the `FormContractor` behavior (https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/master/Builder/FormContractor.php#L110-L111)?

wouterj commented:
I did a bit of searching the the parentDocumnt type is very thightly coupled to Sonata. Maybe we should first create the form types in the TreeBrowserBundle before continuing to use the type for parentDocument.

What you can do is use [`phpcr_document`](http://symfony.com/doc/current/cmf/bundles/phpcr_odm/forms.html#phpcr-document) form type for the time being.

fbourigault commented:
To use the phpcr_document, I need to provide a class option. What should I put here? I looked at code but didn't find where I can get this information!

dbu commented:
hm, good point. we can't know in general, as any document can be a parent, as the tree is not organized by document classes. that is the appeal of the tree widget that it would just show whatever parents we have. with the phpcr_document, we have a dropdown so we can't just flatten the whole tree into it and query the tree for one document type instead.

if we want a quick and dirty solution we could use a text field where you can paste a repository path? phpcr_document is just a fancy way for the same. and then upgrade once tree has a form type. the text field would not be a production ready thing, but would allow us to move on.

wouterj commented:
Doesn't `phpcr_document` return a Document class instead of just the path id?

dbu commented:
yes, phpcr_document is based on the symfony entity form type:
http://symfony.com/doc/current/cmf/bundles/phpcr_odm/forms.html

which means its essentially relational-inspiried and needs a class.

anyways, as wouter is just now adding the tree form type i think we
hopefully can continue here with the tree widget very soon...

fbourigault commented:
So, we are waiting for symfony-cmf/tree-browser-bundle#87 to be resolved?

wouterj commented:
Fyi, I just created https://github.com/symfony-cmf/tree-browser-bundle/pull/114. I think we can quickly merge it and use it in this PR.

dbu commented:
and its merged. thanks wouter!

fbourigault commented:
I wanted to add `egeloen/ckeditor-bundle` to require-dev packages (first to stop my editor complaining about missing CKEditorType), but I don't know which version constraint I should provide. As a better reason to add `egeloen/ckeditor-bundle` to require-dev, I wanted to write some tests for the new FormType but I don't know what worth to be tested. Some advice would be welcome!

wouterj commented:
* In the sandbox, we use [`egeloen/ckeditor-bundle: ^4.0`](https://github.com/symfony-cmf/cmf-sandbox/blob/master/composer.json#L50), let's go with that.
 * As for testing, I would test the happy flow: Just submit the parent, name, title and body and check if the StaticContent document is successfully created (and in a correct way). I would also test each seperate custom form option. So, if I enable CKEditor, does it show me the CKEditor field for body, etc.

fbourigault commented:
@dbu do you know where the `parentDocument` field come from? I can't see it in any model!

wouterj commented:
@fbourigault use `Symfony\Cmf\Bundle\ContentBundle\Doctrine\Phpcr\StaticContent` when using PHPCR. This has `{get/set}ParentDocument()`.

fbourigault commented:
So, should I provide two FormType one for `Model\StaticContent` and one for `Doctrine\Phpcr\StaticContent`?

dbu commented:
> So, should I provide two FormType one for |Model\StaticContent| and one
> for |Doctrine\Phpcr\StaticContent|?

sounds like a plan. the form type for phpcr static content should point
to the model one as base type.

fbourigault commented:
What do you think about a `StaticContentType` for `Model\StaticContent` and a `DoctrinePhpcrStaticContentType` for `Doctrine\Phpcr\StaticContent` that has `StaticContentType` as parent?

dbu commented:
that sounds right, yes.

wouterj commented:
I propose to move `DoctrinePhpcrStaticContentType` to `Doctrine\Phpcr\Form` and call it `StaticContentType` and keep the model one in `Form`.

This way, we group all PHPCR related stuff in one namespace. This makes it easy to for instance extract PHPCR to adapter bundles in the future (a long open discussion).

dbu commented:
i am afraid of the confusion e.g. in autocomplete when you do StaticContentType and get the two choices. as form types do not extend each other, the model one will not be abstract, so people could accidentally be using it. if we split out phpcr things, we will change namespaces anyways, so no having the phpcr form in Forms does not change anything imho.

fbourigault commented:
Autocompleting the wrong type won't show the right fields (or even do an error about missing properties in data class), I'm voting for the `Doctrine\Phpcr\Form\Type\StaticContentType` version.

fbourigault commented:
I also discovered that some admin fields are provided by admin extensions, we have to remind about converting this as Form Extension to get all features working. We will also need to look at sonata admin tabs and groups (`Sonata\AdminBundle\Mapper\BaseGroupedMapper`) but IMHO work should be done on the sonata side to enable anybody to use a per model FormType like we are doing in this PR).

dbu commented:
i see the autocomplete confusion as a developer experience issue. a dev would use StaticContentType and then be confused that it does not work. its hard to spot the difference with the same class name, and the model type won't tell you its the wrong thing for the phpcr document...

but whatever, if wouter and you both prefer that approach, go for it. its not wrong.

about the admin extensions: agreed, we should split our form types into the main things and the specific things, and then in the sonata integration bundle can use those for admin extensions.

fbourigault commented:
Should I create a meta issue in the admin integration bundle to track all the work to do?

wouterj commented:
> Should I create a meta issue in the admin integration bundle to track all the work to do?
 Show all checks

Sounds good to me.

Thanks for going through the rough part of this development, implementing the first split. Things should be easier for the other bundles once we can finish this one.

fbourigault commented:
Using `StaticContentType` for both form types gives me the following error:

```
An exception has been thrown during the rendering of a template ("Unable to render the form because the block names array contains duplicates: "cmf_content_admin_static_content_static_content_static_content_row", "cmf_content_admin_static_content_static_content_row", "cmf_content_admin_static_content_row", "static_content_row", "static_content_row", "form_row".") in SonataAdminBundle:CRUD:base_edit_form.html.twig at line 58.
```

Anyone have an idea?

fbourigault commented:
The new form types changed the field order. I don't think it matters for now as the parentDocument field is still in the General group and we have to work on this when looking at groups/tabs support for sonata integration.

Do you agree?

dbu commented:
this is a 2.0 change so different order seems no big issue to me, as
long as the order makes sense ux wise. if it does not make sense, we
need to check if we can reorder somehow.

fbourigault commented:
@dbu I agree, it should be worked on when admin will be extracted. It will be easier to handle order/groups/tabs in the sonata integration bundle (and I have some ideas).

@wouterj what's the right way to add assets for the `TreeSelectoryType`?

fbourigault commented:
Assets loading is handled by symfony-cmf/tree-browser-bundle#117.

fbourigault commented:
As of writing tests, I found an error in my code about the name field not being in the right StaticContentType. I moved it from model to phpcr type.

@wouterj in a7b61da I added a simple test (based on symfony documentation http://symfony.com/doc/current/form/unit_testing.html). I never tested my form types before this one. Am I correct to only focus on fields added in the tested type?

dbu commented:
> As of writing tests, I found an error in my code about the name field
> not being in the right StaticContentType. I moved it from model to phpcr
> type.

which goes to show that having different class names for the two types
and not only different namespaces might be better for DX. we could still
change that now... (but i shut up now, if you want to keep it, lets keep it)

fbourigault commented:
IIRC I just forgot to move the name field to the phpcr version of `StaticContentType`. Was not a mistake about using the same name. Nice try @dbu ;)

wouterj commented on Tests/Unit/Form/Type/StaticContentTypeTest.php:
```diff
@@ -0,0 +1,92 @@
+        $view = $form->createView();
+        foreach (array_keys($data) as $key) {
+            $this->assertArrayHasKey($key, $view->children);
+        }
```
this doesn't test much, if the key didn't exists, `->submit()` would fail

ElectricMaxxx commented:
can you add the stuff you have extracted into the [sonata-integration](https://github.com/symfony-cmf/sonata-admin-integration-bundle)  and bring that all together into the WIP branch of the sandbox: https://github.com/symfony-cmf/cmf-sandbox/pull/332 later, please? there we will see the results at all.

In my case of the routing bundle it isn't that  easy as the Route Document is devided into subforms accross tabs, so had to think an other way: https://github.com/symfony-cmf/routing-bundle/pull/355/files There i created extra DTOs to fill the subforms, it is still WIP.

fbourigault commented:
This PR is not about moving the Admin class outside of this bundle to the sonata integration but to extract form types and then (in another PR) move the admin class in the sonata integration bundle. By doing this way, there is no need to have a wip branch on the sandbox, but the work can be done graduately.

For the fields ordering, grouping and tabing, I have ideas that will not anyway have to use some objects, the FormTypes will be usable like any FormType (even in a classic Controller), stuff like `PublishableExtension` have to be moved to FormExtension, so the field will be added to the type.

Doing things like this will break ordering and grouping. To fix it, I think we could do something like this:

```php
protected function configureFormFields(FormMapper $formMapper)
{
    $builder = $formFactory->create(StaticContentType::class, null, [
        'inherit_data' => true,
        'label' => false,
        'use_ivory_ckeditor' => !empty($this->ivoryCkeditor),
        'ivory_ckeditor_options' => $this->ivoryCkeditor,
        'parent_document_options' => [
            'widget' => 'compact',
        ],
    ]);

    $formMapper
        ->with('form.group_general')
            ->add($builder->get('parentDocument'))
            -...
        ->end()
        ->with('form.group_routing')
            ->add($builder->get('...'))
            ...
        ->end()
    ;
}
```

With such code, responsibilities are separated and any other admin system (like EasyAdminBundle) could use an other ordering (if supporting).

As writing this comment, I realized that it may be possible in this PR. I will give it a try later today.

@wouterj what do you think of this possible implementation?

fbourigault commented on Tests/Unit/Form/Type/StaticContentTypeTest.php:
```diff
@@ -0,0 +1,92 @@
+        $form->submit($data);
+
+        $object = new StaticContent();
+        $object->setTitle($data['title']);
```
Does starting with an empty `$object` is better? Anyway `$form->submit($data)` will fulfill `title` and `body`!

fbourigault commented:
I fixed field ordering and the solution looks promising if https://github.com/sonata-project/SonataAdminBundle/issues/4100 is confirmed and fixed.

To get this PR fully working, we will also need https://github.com/symfony-cmf/tree-browser-bundle/pull/112 to be fixed.

fbourigault commented on Tests/Unit/Doctrine/Phpcr/Form/Type/StaticContentTypeTest.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        $data = [
+            'name' => 'hello-world',
+            'parentDocument' => '/cms/content',
```
I use a string here as parentDocument, but in docblock I see object. What are the requirements for an object to be used as parent? Is there an interface to implement or it's just the mapping that allows it?

fbourigault commented:
I added the missing `DocumentToPathTransformer` to get the `DocumentToPathTransformer` field working.

However, if I change the parent document and then save, I get the following exception:
```
Doctrine\ODM\PHPCR\PHPCRException: The Id is immutable (/cms/content/team/seoMetadata !== /cms/content/demo/team/seoMetadata). Please use DocumentManager::move to move the document: Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr\SeoMetadata@0000000026d266f0000000001212308f (/cms/content/demo/team/seoMetadata)
```

How should I handle this?

fbourigault commented:
I think this PR is ready for review but still blocked by release of https://github.com/sonata-project/SonataAdminBundle/pull/4101.

wouterj commented on Admin/StaticContentAdmin.php:
```diff
@@ -48,20 +46,27 @@ protected function configureListFields(ListMapper $listMapper)
+            $parent = $formFactory->createNamedBuilder('parentDocument', TextType::class, null, [
+                'disabled' => true,
+            ]);
+            $parent->addModelTransformer(new DocumentToPathTransformer($this->getModelManager()->getDocumentManager()));
```
this always has to be applied, also when creating a new document (then `reverseTransform` is called on form submit)

fbourigault replied:
I just added this here because I don't use the field from StaticContentType

fbourigault commented:
As a side note, what I did with the FormType in the admin, disable all behaviors that may be added to the `StaticContentType` like listeners.

wouterj commented:
This looks very great to me!

@symfony-cmf/core can you please share your opinions/review on this PR? If we all agree on this PR, we'll implement it the same way in all other bundles, so this is quite an important PR.

fbourigault commented:
What do you think if I move the read-only parent document stuff into the type and add an option to set it read-only or not? Any StaticContentType client have to handle move by itself, so it could be good.

fbourigault commented:
Readonly parentDocument is now handled in `Doctrine\Phpcr\StaticContentType`.

dbu commented on Doctrine/Phpcr/Form/Type/StaticContentType.php:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
we should document the supported values in $options. we should also use the OptionsResolver to make sure the options are not all required. we can then also validate those options to only have valid values. this will make it easier to discover mistakes with the form builder.

dbu replied:
ah sorry, this is done by the form system. still, a doc here that lists the options and explains what they do would be useful

fbourigault replied:
How would you describe such options in docblock?

dbu replied:
maybe similar to this:
https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/Varnish.php#L25-L27

note that the "additional" comes because in that example there is a base
class and we add to it. i feel `-` is better to indicate the list than
`*` as `*` is already used to denote the comment section.

dbu commented on Doctrine/Phpcr/Form/Type/StaticContentType.php:
```diff
@@ -0,0 +1,95 @@
+        }
+
+        $builder->get('parentDocument')
+            ->addModelTransformer(new DocumentToPathTransformer($this->documentManager))
```
is this also needed when using the tree select type, or only with the text type?

fbourigault replied:
This is needed for both to get the `parentDocument` path as string.

dbu replied:
@wouterj is this the intended behaviour of TreeSelectType? should it not work with documents as well?

dbu commented on Form/Type/StaticContentType.php:
```diff
@@ -0,0 +1,58 @@
+            'data_class' => StaticContent::class,
+            'use_ivory_ckeditor' => false,
+            'ivory_ckeditor_options' => [],
+        ]);
```
we could add validation to make use_ivory_ckeditor boolean, and ivory_ckeditor_options an array

dbu commented:
this looks good to me, great job! i added a few details, the main thing is that i think using the validation of OptionsResolver could further improve detecting mistakes when configuring, and a bit of phpdoc would help us to write the symfony-cmf-docs.

speaking of which, i think updating the symfony-cmf-docs is something we really should start in parallel.

fbourigault commented:
I added the validation and phpdoc.

dbu commented on Form/Type/StaticContentType.php:
```diff
@@ -0,0 +1,66 @@
+            ->add(
+                'body',
+                $options['use_ivory_ckeditor'] ? CKEditorType::class : TextareaType::class,
+                $options['ivory_ckeditor_options']
```
that question seems to have disappeared: does it make sense to pass the options to the text area as well? imho we should either call this editor_options and explain its for ckeditor or textarea, depending on the use_ivory_ckeditor flag, or have different options fields for the text area (or no options if it makes no sense to pass options to text area)

fbourigault replied:
Does it make sense to allow customization of the `TextareaType`? Looks like there are no useful options. Anyway, I was wondering about using a FormExtension automatically enabled if IvoryCKEditorBundle is available. This will move the weird IvoryCKEditor options to it's own class. What do you think?

dbu replied:
if we can simplify this, that would be great. FormExtension sounds like
it could be more flexible and open the full power to the user instead of
us trying to forward some flags to another part...

how would the admin look if we use a FormExtension? (or more important:
would it make it harder for the user to set up ivory?)

fbourigault replied:
In https://github.com/symfony-cmf/content-bundle/blob/master/DependencyInjection/CmfContentExtension.php#L40-L66, we could instead conditionally register the FormExtension.

So the user will still setup ivory through the ContentBundle configuration (https://github.com/symfony-cmf/content-bundle/blob/master/DependencyInjection/Configuration.php#L45-L57).

With such setup, the admin will have no knowledge (and don't need) about type used for the body field.

dbu commented:
looks good to me!

@wouterj can you also help think through the options again? as we add something new anyways, we should think how we  want to name such options and how to build that. like is it use_ivory_ckeditor, or rather just editor with values ivory_ckeditor and textarea, to make it easy to add more options?

fbourigault commented:
Latest commit introduce the FormExtension for the CKEditorType. I really like it! If it's ok, I will add a tests for the extension.

ElectricMaxxx commented on Resources/config/forms-phpcr.xml:
```diff
@@ -0,0 +1,15 @@
+
+            <argument type="service" id="doctrine_phpcr.odm.document_manager" />
+            <argument>%cmf_content.persistence.phpcr.content_basepath%</argument>
+            <tag name="form.type" />
```
do we still need that tag where we have the `::` Type stuff? I heard we need the service definition only.

fbourigault replied:
As the `FormType` has dependencies, we still need to register it as as service and so need to tag the service.

ElectricMaxxx replied:
ah right, it was the `alias` we do not need anymore.

ElectricMaxxx commented:
It is really great to see it growing. But i do have one question at all. Is that way of defining form types for the whole document/model and separating it through soanata's form-builder into the places we need each field compatible with our plan to use the form outside of sonata too? Can we use that form-builder to create a simple, but different, custom form or custom admin with a different order of the fields? Asking that cause we have other admins where the properties of a document are separated into diferent tabs means different form.

fbourigault commented:
As Symfony does not provide fields ordering/grouping/tabbings support natively, each admin will use a different system. Sonata use the FormMapper, but any other admin could use any system and within a standard controller/view, you will control fields ordering in the template using calls to form_row.

Do you have an example of admin having different forms per tab?

ElectricMaxxx commented:
@fbourigault doing same for routing-bundle would be really easy. That would be an admin with a really fragmented admin form with one model/document as the base. So if we do so, as you said, the sonata extraction (incl. providing own form types) will be a lot more easy :-)

fbourigault commented:
I didn't yet looked a lot at routing-bundle but I think it will be easier. Only block-bundle looks to be harder as there are a lot of admins!

ElectricMaxxx commented:
Would it be hard to create an own Independent Form builder. Doing so we would provide a really flexible solution. A user can take our form builder and our form types an put them whereever he wants. Normally i would expect such a builder inside form component to re-arage Form without doing fancy templating.

fbourigault commented:
I don't know if it will be hard, but I think it's not a good thing to provide our own FormBuilder as it will not be Symfony standard. I prefer keep the solution chooses here and wait for something implemented withing the Symfony\Form component. This PR is a great forward move to reusability of content-bundle but IMHO we must not introduce some new behaviora that let this PR being a backward move. What so you think @wouterj & @dbu ?

dbu commented:
i think the forms are an improvement over forms baked into sonata admin.
i don't see how we could make this more flexible. providing a form type
for a title field or the body field becomes pointless, then you can just
build your own form. or use the content-type bundle dan is building and
let it generate a form.

lets wrap this form extracting up (and its not just extraction, also
modernization to make the forms work without sonata widgets) and merge.

before we tackle all the other bundles, we should discuss if there is
another strategy. not that we move in the wrong direction or have to
redo everything. but lets move that discussion to its own issue if you
want to follow up on it. but imho you either need custom templates
anyways and then control the order, or use an admin generator like
sonata that can reorder. for custom form builders, there is also
https://github.com/egeloen/IvoryOrderedFormBundle btw.

dbu commented on Form/Extension/CKEditorExtension.php:
```diff
@@ -0,0 +1,49 @@
+use Symfony\Component\Form\AbstractTypeExtension;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class CKEditorExtension extends AbstractTypeExtension
```
lets consistently use the Ivory prefix (for the class, and also for the service xml file). the extension is specific to the ivory thing, not general ckeditor...

dbu commented on Form/Extension/CKEditorExtension.php:
```diff
@@ -0,0 +1,49 @@
+    private $options;
+
+    /**
+     * @param array $options CKEditorType options
```
lets add a line `@see CKEditorType for documentation on the options` to make it really clear

dbu commented on Resources/config/ckeditor.xml:
```diff
@@ -0,0 +1,14 @@
+
+    <services>
+        <service id="cmf_content.form.extension.ckeditor"
+            class="Symfony\Cmf\Bundle\ContentBundle\Form\Extension\CKEditorExtension">
```
should we add something to configure editor options in the content bundle? i would say that would be useful, kind of as we had before, only not looped through sonata admin.

fbourigault replied:
You mean through the parameter I removed?

dbu replied:
CKEditorExtension accepts options in the constructor. i think we still have the options in the Configuration.php but should pass them to the extension here.

fbourigault replied:
In the Configuration, we only allow to pass the config_name as option (https://github.com/symfony-cmf/content-bundle/blob/master/DependencyInjection/CmfContentExtension.php#L63-L65). As it uses a parameter, anyone could override it. I can reintroduce the same code and use this parameter as extension argument.

dbu commented:
hey, that form extension is so much better than the old approach! love it!

(and yes, as before there will be only one set of configurations for the body - if anybody needs different editors for different bodies somehow (extending documents or something) then they just need to use the ckeditor type directly and pass options)

dbu commented:
thanks for the updates. this looks ready to me. if you can add a test for the ivory form extension, thats great.

i think we don't need to rewrite the history too much. simply squashing all commits seems good enough, to have one clean commit to merge.

@wouterj do you want to look over this again before we hit merge?

fbourigault commented:
I started to write a test but didn't finished. I will also squash all 7 commits to a single one and this will be mergeable :)

fbourigault commented:
I need help for tests. I started to write the `IvoryCKEditorExtension` tests using a `TypeTestCase` but the `TypeTestCase::getExtensions` method has to return an array of `FormExtensionInterface`. As my `IvoryCKEditorExtension` is a  `FormTypeExtensionInterface`, I wanted to write a `FormExtensionInterface` but I don't know where to store it. If I look at the `CsrfFormExtension`, it's stored inside the `src` folder but only used for tests. I want to store it in the `Tests` folder but I don't know where. Anyone have an idea?

dbu commented:
if you only need it in a single test case and its a very simple class, you could add the class below the test case in the same file. not the most elegant way, but for testing its good enough. if that feels wrong to you, i'd put it in Tests/Resources/Form

fbourigault commented:
@dbu when writing tests, I found that `Symfony\Component\Form\PreloadedExtension` can load my `IvoryCKEditorExtension` :)

dbu commented:
ah, even better :-)

the build is green too \o/

is this ready to merge now, or anything left to do?

fbourigault commented:
squashing commits!

dbu commented:
okay, i guess its time for doing that then :-)

@wouterj can you also check if anything is missing?

fbourigault commented:
commits are now squashed.

fbourigault commented:
@wouterj how does it looks to you?

dbu commented:
thanks a lot!

@wouterj if you see something, lets just fix that in a follow-up PR. i merged this one so we can move on.
